### PR TITLE
actool: new subcommands: cat-manifest and patch-manifest

### PR DESCRIPTION
**Pull request by alban**
The manifest can be printed as is: 
```
  $ actool cat-manifest file.aci
```

Or it can be parsed and printed in a pretty form:
```
  $ actool cat-manifest --pretty-print file.aci
```

The ACI can be copied with an alternative manifest:
```
  $ actool patch-manifest --manifest=./manifest file.aci target.aci
  $ mv target.aci file.aci
```

Or this can be done in one operation with --replace:
```
  $ actool patch-manifest --manifest=./manifest --replace file.aci
```

The manifest could be patched without providing the whole manifest file:
```
  $ actool patch-manifest --name=example.com/app \
                          --capability=CAP_NET_ADMIN,CAP_SYS_ADMIN \
                          --replace \
                          file.aci
```

### CODE REVIEW

**Comment by alban** [comment]
I updated my branch to also support
```
--mounts=tmp,path=/tmp:tmp2,path=/tmp2,readOnly=true:tmp3,path=/tmp3,readOnly=false
```

**Comment by alban** [comment]
Strace reveals that the aci file is written by chunks of 64 bytes and it is quite slow for big images. I have not found how to increase the buffer of the gzip writer.

If increasing the write buffer is not enough, maybe pgzip can help: https://github.com/appc/spec/issues/259

File: actool/manifest.go [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+		if exit != 0 && !patchOverwrite {
+			os.Remove(fh.Name())
+		}
+	}()
+
+	// Prepare input reader
+
+	input, err := os.Open(inputFile)
+	if err != nil {
+		stderr("patch-manifest: cannot open %s: v", inputFile, err)
```
Comment (jonboulle) [inline actool/manifest.go]: s/v/%v/

File: actool/manifest.go [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+		return 1
+	}
+	defer input.Close()
+
+	tr, err := aci.NewCompressedTarReader(input)
+	if err != nil {
+		stderr("patch-manifest: cannot extract %s: v", inputFile, err)
```
Comment (jonboulle) [inline actool/manifest.go]: s/v/%v/
...etc

**Comment by jonboulle** [comment]
@alban that's annoying. what about uncompressed?

**Comment by alban** [comment]
@jonboulle When using ``actool patch-manifest --no-compression``, it takes 2 seconds. Quite fast for a big image. A lot of the write() calls write 32768 bytes at once, but still 80% of the writes are below 4k. So the small writes might not be the bottleneck.

With the default compression, it takes 32 seconds, >99% of the writes are 64 bytes and only 5 IOs are >16k. But ``actool`` takes all my CPU so the small IOs might not be the problem. ``perf top`` shows the following:
```
  36.42%  actool                              [.] compress/flate.(*compressor).findMatch
  12.73%  actool                              [.] compress/flate.(*compressor).deflate
```

**Comment by alban** [comment]
When using the BestSpeed option, gzip is faster with my big image (14 seconds) but it is still 7x slower than no compression:
```
-               gw = gzip.NewWriter(fh)
+               gw, _ = gzip.NewWriterLevel(fh, gzip.BestSpeed)
```

/bin/gzip and pgzip both takes 13 seconds.

**Comment by philips** [comment]
@alban we already need to patch gzip to do gzran, maybe just bite the bullet and do it there? https://github.com/coreos/gzran

**Comment by alban** [comment]
@philips : Where is gzran used at the moment? I'm not sure how gzran can help: the slow part is compressing the archive, not uncompressing.

@jonboulle : can the gzip optimizations be done after this gets merged?

File: actool/manifest.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+			if err != nil {
+				return fmt.Errorf("cannot parse mount point %q", m)
+			}
+			app.MountPoints = append(app.MountPoints, *mountPoint)
+		}
+	}
+	return nil
+}
+
+// iterate iterates over the tar reader and optionally write in the tar writer if not nil
```
Comment (jonboulle) [inline actool/manifest.go]: can you call this extractManifest or something and update the docstring pls? "iterate" is a bit vague/misleading

File: actool/manifest.go [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+					return err
+				}
+			} else if tw != nil {
+				err := tw.WriteHeader(hdr)
+				if err != nil {
+					return err
+				}
+				_, err = io.Copy(tw, tr)
+			}
+			if err != nil {
```
Comment (jonboulle) [inline actool/manifest.go]: inconsistency nit, weird to handle error here for the second branch but on line 188 for the first branch

File: actool/manifest.go [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+				return err
+			}
+		default:
+			return fmt.Errorf("error reading tarball: %v", err)
+		}
+	}
+
+	return nil
+
```
Comment (jonboulle) [inline actool/manifest.go]: remove

File: schema/types/mountpoint.go [hunk 6]
```diff
@@ -1,7 +1,57 @@
...
 	ReadOnly bool   `json:"readOnly,omitempty"`
 }
+
+// MountPointFromString takes a command line mountpoint parameter and returns a mountpoint
+//
+// It is useful for actool patch-manifest --mounts
+//
+// Example mountpoint parameters:
+// 	database,path=/tmp,readOnly=true
+func MountPointFromString(mp string) (*MountPoint, error) {
```
Comment (jonboulle) [inline schema/types/mountpoint.go]: please add a test table for this? I haven't been hardline enough about people adding tests with schema code..

File: actool/manifest.go [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+	catPrettyPrint bool
+
+	cmdPatchManifest = &Command{
+		Name:        "patch-manifest",
+		Description: `Copy an ACI and patch its manifest.`,
+		Summary:     "Copy an ACI and patch its manifest (experimental)",
+		Usage: `
+		  [--manifest=MANIFEST_FILE]
+		  [--name=example.com/app]
+                  [--capability=CAP_SYS_ADMIN,CAP_NET_ADMIN]
```
Comment (jonboulle) [inline actool/manifest.go]: indentation is slightly off

File: actool/manifest.go [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+			return fmt.Errorf("no app in the manifest")
+		}
+		isolator := app.Isolators.GetByName(types.LinuxCapabilitiesRetainSetName)
+		if isolator != nil {
+			return fmt.Errorf("isolator already exists")
+		}
+
+		capsList := strings.Split(patchCaps, ",")
+		caps := fmt.Sprintf(`"set": ["%s"]`,
+			strings.Join(capsList, `", "`))
```
Comment (jonboulle) [inline actool/manifest.go]: we don't normally linewrap this aggressively do we?

File: actool/manifest.go [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+		isolatorStr := getIsolatorStr(types.LinuxCapabilitiesRetainSetName,
+			caps)
```
Comment (jonboulle) [inline actool/manifest.go]: likewise

File: actool/manifest.go [hunk 10]
```diff
@@ -0,0 +1,364 @@
...
+		isolator = &types.Isolator{}
```
Comment (jonboulle) [inline actool/manifest.go]: wait.. wuh? why are you creating a JSON and then unmarshalling, what about just instantiating the Isolator directly?

File: actool/manifest.go [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+
+	tr, err := aci.NewCompressedTarReader(input)
+	if err != nil {
+		stderr("cat-manifest: cannot extract %s: %v", inputFile, err)
+		return 1
+	}
+
+	err = iterate(tr, nil, true, nil)
+	if err != nil {
+		stderr("cat-manifest: Unable to read %s: %v", inputFile, err)
```
Comment (jonboulle) [inline actool/manifest.go]: nit, inconsistent capitalisation, please standardise it across the output
`cat-manifest: cannot extract`
`cat-manifest: Unable to read`

File: actool/manifest.go (hunk 10 above)
Comment (alban) [inline actool/manifest.go]: The ``Isolator`` struct is defined in a different package ("types") and the ``IsolatorValue`` field is not exported. The only method initializing the private field is ``UnmarshalJSON``.

The ``IsolatorValue`` for ``LinuxCapabilitiesRetainSet`` also contain a private field without method or constructor to initialize it, except ``UnmarshalJSON``.

I added https://github.com/appc/spec/issues/268 and I will put a TODO comment with a link to it.

File: schema/types/mountpoint.go (hunk 6 above)
Comment (alban) [inline schema/types/mountpoint.go]: Added in ``schema/types/mountpoint_test.go``

**Comment by alban** [comment]
@jonboulle : thanks for the review! I updated my branch.

**Comment by alban** [comment]
I added --user and --group. That should help for testing https://github.com/coreos/rocket/issues/539

File: actool/manifest.go (hunk 10 above)
Comment (jonboulle) [inline actool/manifest.go]: egads, sorry about that.

File: actool/manifest.go [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+		Run:         runCatManifest,
+	}
+)
+
+func init() {
+	cmdPatchManifest.Flags.BoolVar(&patchOverwrite, "overwrite", false, "Overwrite target file if it already exists")
+	cmdPatchManifest.Flags.BoolVar(&patchNocompress, "no-compression", false, "Do not gzip-compress the produced ACI")
+	cmdPatchManifest.Flags.BoolVar(&patchReplace, "replace", false, "Replace the input file")
+
+	cmdPatchManifest.Flags.StringVar(&patchManifestFile, "manifest", "", "Manifest file")
```
Comment (jonboulle) [inline actool/manifest.go]: a bit ambiguous, prefer something like `Replace image manifest with this file. Incompatible with other replace options.`

**Comment by jonboulle** [comment]
one tiny thing but LGTM after that

File: actool/manifest.go (hunk 12 above)
Comment (alban) [inline actool/manifest.go]: I agree, thanks for the phrasing.

**Comment by alban** [comment]
Branch updated.

**Comment by alban** [comment]
I added `--exec` (/cc @krnowak ) and rebased the branch.

**Comment by jonboulle** [comment]
LGTM, thanks, this is great!

**Pull request merged**